# PANDAS PROJECT

# 0. Environment

### Import Modules

In [1]:
import pandas as pd
import numpy as np
import pandas_profiling
import re
pd.set_option('display.max_columns',None)
pd.set_option('display.max_colwidth', 200)

### Import CSV

In [2]:
data = pd.read_csv('GSAF5.csv',encoding='iso8859-1')

# 1. Checking the DataFrame 

## 1.1 General look of the DataFrame

In [3]:
data.head()

Case Number       Date  Year        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.15  16-Sep-16  2016  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex   Age  \
0  New Smyrna Beach, Volusia County  Surfing             male    M   16   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano    M   36   
2  New Smyrna Beach, Volusia County  Surfing             male    M   43   
3                  Thirteenth Beach  Surfing  Rory Angiolella    M  NaN   
4                       Bells Beach  Surfing             male    M  NaN   

                                  Injury Fatal (Y/N)   Time   Species   \
0                  Minor injury to thigh           N  13h00        NaN   
1                   Lacerations to hands           N  11h00        NaN   
2               Lacerations to lower leg           N  10h43        NaN   
3           Struck by fin on chest & leg           N    NaN        NaN   
4  No injury: Knocked off board by shark           N    NaN  2 m shark   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                                                      href formula  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

                                                                              href  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

  Case Number.1 Case Number.2  original order Unnamed: 22 Unnamed: 23  
0  2016.09.18.c  2016.09.18.c            5993         NaN         NaN  
1  2016.09.18.b  2016.09.18.b            5992         NaN         NaN  
2  2016.09.18.a  2016.09.18.a            5991         NaN         NaN  
3    2016.09.17    2016.09.17            5990         NaN         NaN  
4    2016.09.16    2016.09.15            5989         NaN         NaN

In [4]:
data.describe().T

count         mean          std  min      25%     50%  \
Year            5992.0  1925.204606   286.473712  0.0  1942.00  1975.0   
original order  5992.0  2997.548899  1729.860210  2.0  1499.75  2997.5   

                    75%     max  
Year            2003.00  2016.0  
original order  4495.25  5993.0

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5992 entries, 0 to 5991
Data columns (total 24 columns):
Case Number               5992 non-null object
Date                      5992 non-null object
Year                      5992 non-null int64
Type                      5992 non-null object
Country                   5949 non-null object
Area                      5590 non-null object
Location                  5496 non-null object
Activity                  5465 non-null object
Name                      5792 non-null object
Sex                       5425 non-null object
Age                       3311 non-null object
Injury                    5965 non-null object
Fatal (Y/N)               5973 non-null object
Time                      2779 non-null object
Species                   3058 non-null object
Investigator or Source    5977 non-null object
pdf                       5992 non-null object
href formula              5991 non-null object
href                      5989 non-null object
C

In [6]:
data.shape

(5992, 24)

In [7]:
data.dtypes

Case Number               object
Date                      object
Year                       int64
Type                      object
Country                   object
Area                      object
Location                  object
Activity                  object
Name                      object
Sex                       object
Age                       object
Injury                    object
Fatal (Y/N)               object
Time                      object
Species                   object
Investigator or Source    object
pdf                       object
href formula              object
href                      object
Case Number.1             object
Case Number.2             object
original order             int64
Unnamed: 22               object
Unnamed: 23               object
dtype: object

In [8]:
#data.profile_report()

In [9]:
data.columns

Index(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex ', 'Age', 'Injury', 'Fatal (Y/N)', 'Time',
       'Species ', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23'],
      dtype='object')

**The column Species and Sex have spaces at the end on the name, we are going to take them out**

In [10]:
data.columns = ['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Location',
       'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal', 'Time',
       'Species', 'Investigator or Source', 'pdf', 'href formula', 'href',
       'Case Number.1', 'Case Number.2', 'original order', 'Unnamed: 22',
       'Unnamed: 23']

**Reorder columns to work more easily**

In [11]:
data.head(3)

Case Number       Date  Year        Type Country     Area  \
0  2016.09.18.c  18-Sep-16  2016  Unprovoked     USA  Florida   
1  2016.09.18.b  18-Sep-16  2016  Unprovoked     USA  Florida   
2  2016.09.18.a  18-Sep-16  2016  Unprovoked     USA  Florida   

                           Location Activity            Name Sex Age  \
0  New Smyrna Beach, Volusia County  Surfing            male   M  16   
1  New Smyrna Beach, Volusia County  Surfing  Chucky Luciano   M  36   
2  New Smyrna Beach, Volusia County  Surfing            male   M  43   

                     Injury Fatal   Time Species       Investigator or Source  \
0     Minor injury to thigh     N  13h00     NaN  Orlando Sentinel, 9/19/2016   
1      Lacerations to hands     N  11h00     NaN  Orlando Sentinel, 9/19/2016   
2  Lacerations to lower leg     N  10h43     NaN  Orlando Sentinel, 9/19/2016   

                        pdf  \
0      2016.09.18.c-NSB.pdf   
1  2016.09.18.b-Luciano.pdf   
2      2016.09.18.a-NSB.pdf   

                                                                     href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2      http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   

                                                                             href  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2      http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   

  Case Number.1 Case Number.2  original order Unnamed: 22 Unnamed: 23  
0  2016.09.18.c  2016.09.18.c            5993         NaN         NaN  
1  2016.09.18.b  2016.09.18.b            5992         NaN         NaN  
2  2016.09.18.a  2016.09.18.a            5991         NaN         NaN

In [12]:
data = data[['Case Number', 'Case Number.1', 'Case Number.2', 'Date', 'Year','Time',
             'Type', 'Country', 'Area', 'Location', 'Activity', 'Name', 'Sex', 'Age',
             'Injury', 'Fatal','Species', 'Investigator or Source', 'pdf',
             'href formula', 'href', 'original order', 'Unnamed: 22', 'Unnamed: 23']]

In [13]:
data.head(3)

Case Number Case Number.1 Case Number.2       Date  Year   Time  \
0  2016.09.18.c  2016.09.18.c  2016.09.18.c  18-Sep-16  2016  13h00   
1  2016.09.18.b  2016.09.18.b  2016.09.18.b  18-Sep-16  2016  11h00   
2  2016.09.18.a  2016.09.18.a  2016.09.18.a  18-Sep-16  2016  10h43   

         Type Country     Area                          Location Activity  \
0  Unprovoked     USA  Florida  New Smyrna Beach, Volusia County  Surfing   
1  Unprovoked     USA  Florida  New Smyrna Beach, Volusia County  Surfing   
2  Unprovoked     USA  Florida  New Smyrna Beach, Volusia County  Surfing   

             Name Sex Age                    Injury Fatal Species  \
0            male   M  16     Minor injury to thigh     N     NaN   
1  Chucky Luciano   M  36      Lacerations to hands     N     NaN   
2            male   M  43  Lacerations to lower leg     N     NaN   

        Investigator or Source                       pdf  \
0  Orlando Sentinel, 9/19/2016      2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016  2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016      2016.09.18.a-NSB.pdf   

                                                                     href formula  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2      http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   

                                                                             href  \
0      http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2      http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   

   original order Unnamed: 22 Unnamed: 23  
0            5993         NaN         NaN  
1            5992         NaN         NaN  
2            5991         NaN         NaN

# 2. CLEANING THE DATA

## 2.1. We start with the columns

### 2.1.1. Set the index

**The column Original Order 
seems to be the old index. As we will keep  the index of the dataframe, we will delete it**

In [14]:
data = data.drop(['original order'], axis=1)

### 2.1.2. Check if there are empty columns

**Checking the columns with more than 2000 nulls out of 5992**

In [15]:
null_cols = data.isnull().sum()
null_cols[null_cols > 2000]

Time           3213
Age            2681
Species        2934
Unnamed: 22    5991
Unnamed: 23    5990
dtype: int64

**The cols 'Unnamed: 22' and 'Unnamed: 23' have only one value and two values.
We check the values.**

In [16]:
unnamed22 = data['Unnamed: 22'].unique()
unnamed22

array([nan, 'stopped here'], dtype=object)

In [17]:
unnamed23 = data['Unnamed: 23'].unique()
unnamed23

array([nan, 'Teramo', 'change filename'], dtype=object)

**We check the values of the rows of those values.**

In [18]:
data[data['Unnamed: 22']=='stopped here']

Case Number Case Number.1 Case Number.2       Date  Year   Time  \
1247  2006.06.18    2006.06.18    2006.06.18  18-Jun-06  2006  09h00   

            Type Country        Area                        Location  \
1247  Unprovoked  BRAZIL  Pernambuco  Punta Del Chifre Beach, Olinda   

           Activity                     Name Sex Age                   Injury  \
1247  Body boarding  Humberto Pessoa Batista   M  27  Left thigh bitten FATAL   

     Fatal Species Investigator or Source                     pdf  \
1247     Y     NaN     globalsurfnews.com  2006.06.18-Batista.pdf   

                                                                      href formula  \
1247  http://sharkattackfile.net/spreadsheets/pdf_directory/2006.06.18-Batista.pdf   

                                                                              href  \
1247  http://sharkattackfile.net/spreadsheets/pdf_directory/2006.06.18-Batista.pdf   

       Unnamed: 22 Unnamed: 23  
1247  stopped here         NaN

In [19]:
data[data['Unnamed: 23']=='Teramo']

Case Number Case Number.1 Case Number.2       Date  Year Time      Type  \
4182  1952.07.13    1952.07.13    1952.07.13  13-Jul-52  1952  NaN  Provoked   

     Country        Area                     Location Activity  \
4182     USA  California  San Diego, San Diego County  Fishing   

                                                     Name Sex Age  \
4182  Gerald Howard, on board sportsfishing boat Teresa A   M  34   

                                                             Injury Fatal  \
4182  Part of hand removed by shark he had caught PROVOKED INCIDENT     N   

     Species  Investigator or Source                    pdf  \
4182     NaN  L.A. Times, 7/14/1952   1952.07.13-Howard.pdf   

                                                                     href formula  \
4182  http://sharkattackfile.net/spreadsheets/pdf_directory/1952.07.13-Howard.pdf   

                                                                             href  \
4182  http://sharkattackfile.net/spreadsheets/pdf_directory/1952.07.13-Howard.pdf   

     Unnamed: 22 Unnamed: 23  
4182         NaN      Teramo

In [20]:
data[data['Unnamed: 23']=='change filename']

Case Number Case Number.1 Case Number.2         Date  Year Time     Type  \
5585  1879.03.10    1879.03.10    1879.03.10  10-Mar-1879  1879  NaN  Invalid   

        Country             Area      Location  \
5585  AUSTRALIA  New South Wales  Near Sydney    

                                           Activity  \
5585  The steamship Bonnie Dundee lost in collision   

                                Name Sex  Age  \
5585  Cabin boy of the Bonnie Dundee   M  NaN   

                               Injury Fatal Species Investigator or Source  \
5585  Partial  remains found in shark     Y     NaN        Star, 3/22/1879   

                               pdf  \
5585  1879.03.10-Bonnie-Dundee.pdf   

                                                                            href formula  \
5585  http://sharkattackfile.net/spreadsheets/pdf_directory/1879.03.10-Bonnie-Dundee.pdf   

                                                                                    href  \
5585  http://sharkattackfile.net/spreadsheets/pdf_directory/1879.03.10-Bonnie-Dundee.pdf   

     Unnamed: 22      Unnamed: 23  
5585         NaN  change filename

**As the info in the cell seems no relevant, so we drop the columns 'Unnamed: 22' and 'Unnamed: 23'.**

In [21]:
data = data.drop(['Unnamed: 22', 'Unnamed: 23'], axis=1)

In [22]:
data.head()

Case Number Case Number.1 Case Number.2       Date  Year   Time  \
0  2016.09.18.c  2016.09.18.c  2016.09.18.c  18-Sep-16  2016  13h00   
1  2016.09.18.b  2016.09.18.b  2016.09.18.b  18-Sep-16  2016  11h00   
2  2016.09.18.a  2016.09.18.a  2016.09.18.a  18-Sep-16  2016  10h43   
3    2016.09.17    2016.09.17    2016.09.17  17-Sep-16  2016    NaN   
4    2016.09.15    2016.09.16    2016.09.15  16-Sep-16  2016    NaN   

         Type    Country      Area                          Location Activity  \
0  Unprovoked        USA   Florida  New Smyrna Beach, Volusia County  Surfing   
1  Unprovoked        USA   Florida  New Smyrna Beach, Volusia County  Surfing   
2  Unprovoked        USA   Florida  New Smyrna Beach, Volusia County  Surfing   
3  Unprovoked  AUSTRALIA  Victoria                  Thirteenth Beach  Surfing   
4  Unprovoked  AUSTRALIA  Victoria                       Bells Beach  Surfing   

              Name Sex  Age                                 Injury Fatal  \
0             male   M   16                  Minor injury to thigh     N   
1   Chucky Luciano   M   36                   Lacerations to hands     N   
2             male   M   43               Lacerations to lower leg     N   
3  Rory Angiolella   M  NaN           Struck by fin on chest & leg     N   
4             male   M  NaN  No injury: Knocked off board by shark     N   

     Species       Investigator or Source                        pdf  \
0        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1        NaN  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2        NaN  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3        NaN           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4  2 m shark           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                                                      href formula  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

                                                                              href  
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf  
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf  
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf  
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf  
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf

### 2.1.3 Check if there are duplicate columns to delete them

**We will check if there is any duplicated comlumn**

In [23]:
#source: https://thispointer.com/how-to-find-drop-duplicate-columns-in-a-dataframe-python-pandas/
def getDuplicateColumns(df):
    duplicateColumnNames = set()
    # Iterate over all the columns in dataframe
    for x in range(df.shape[1]):
        # Select column at xth index.
        col = df.iloc[:, x]
        # Iterate over all the columns in DataFrame from (x+1)th index till end
        for y in range(x + 1, df.shape[1]):
            # Select column at yth index.
            otherCol = df.iloc[:, y]
            # Check if two columns at x 7 y index are equal
            if col.equals(otherCol):
                duplicateColumnNames.add(df.columns.values[y])
 
    return list(duplicateColumnNames)

In [24]:
getDuplicateColumns(data)

[]

**There is no complitely duplicate columns.**

**Let's check if with a function we can found possible very similar columns.**

**Create a function to detect percentage of similitud amog 2 rows**

In [25]:
def compareRows(df, col1, col2):
    eq_row = np.where(df[col1] == df[col2], 1, 0).sum()
    rows = df.shape[0] - df[col1].isnull().sum()
    per_row = round((eq_row / rows) * 100,2)
    return per_row

**Make a table comparing column to column**

In [26]:
compared_cols = pd.DataFrame()
compared_cols['index'] = data.columns
compared_cols = compared_cols.set_index('index')
for col in data.columns:
    column = list()
    for col2 in data.columns:
        column.append(compareRows(data, col, col2))
    compared_cols[col] = column
 

In [27]:
compared_cols

Case Number  Case Number.1  Case Number.2    Date  \
index                                                                       
Case Number                  100.00          99.78          99.97    0.05   
Case Number.1                 99.78         100.00          99.82    0.05   
Case Number.2                 99.97          99.82         100.00    0.05   
Date                           0.05           0.05           0.05  100.00   
Year                           0.00           0.00           0.00    0.00   
Time                           0.00           0.00           0.00    0.00   
Type                           0.00           0.00           0.00    0.00   
Country                        0.00           0.00           0.00    0.00   
Area                           0.00           0.00           0.00    0.00   
Location                       0.00           0.00           0.00    0.00   
Activity                       0.00           0.00           0.00    0.00   
Name                           0.00           0.00           0.00    0.00   
Sex                            0.00           0.00           0.00    0.00   
Age                            0.00           0.00           0.00    0.00   
Injury                         0.00           0.00           0.00    0.00   
Fatal                          0.00           0.00           0.00    0.00   
Species                        0.00           0.00           0.00    0.00   
Investigator or Source         0.00           0.00           0.00    0.00   
pdf                            0.02           0.02           0.02    0.00   
href formula                   0.00           0.00           0.00    0.00   
href                           0.00           0.00           0.00    0.00   

                         Year    Time    Type  Country    Area  Location  \
index                                                                      
Case Number               0.0    0.00    0.00      0.0    0.00      0.00   
Case Number.1             0.0    0.00    0.00      0.0    0.00      0.00   
Case Number.2             0.0    0.00    0.00      0.0    0.00      0.00   
Date                      0.0    0.00    0.00      0.0    0.00      0.00   
Year                    100.0    0.00    0.00      0.0    0.00      0.00   
Time                      0.0  100.00    0.00      0.0    0.00      0.00   
Type                      0.0    0.00  100.00      0.0    0.00      0.00   
Country                   0.0    0.00    0.00    100.0    0.00      0.00   
Area                      0.0    0.00    0.00      0.0  100.00      0.13   
Location                  0.0    0.00    0.00      0.0    0.13    100.00   
Activity                  0.0    0.00    0.23      0.0    0.00      0.00   
Name                      0.0    0.00    0.00      0.0    0.00      0.00   
Sex                       0.0    0.00    0.00      0.0    0.00      0.00   
Age                       0.0    0.00    0.00      0.0    0.00      0.00   
Injury                    0.0    0.04    0.00      0.0    0.00      0.00   
Fatal                     0.0    0.00    0.00      0.0    0.00      0.00   
Species                   0.0    0.00    0.00      0.0    0.00      0.00   
Investigator or Source    0.0    0.00    0.00      0.0    0.00      0.00   
pdf                       0.0    0.00    0.00      0.0    0.00      0.00   
href formula              0.0    0.00    0.00      0.0    0.00      0.00   
href                      0.0    0.00    0.00      0.0    0.00      0.00   

                        Activity    Name     Sex    Age  Injury   Fatal  \
index                                                                     
Case Number                 0.00    0.00    0.00    0.0    0.00    0.00   
Case Number.1               0.00    0.00    0.00    0.0    0.00    0.00   
Case Number.2               0.00    0.00    0.00    0.0    0.00    0.00   
Date                        0.00    0.00    0.00    0.0    0.00    0.00   
Year                        0.00    0.00    0.00    0

**Case number is very similar to Case Number.1 and Case Sumber.2**
**href and href formula are very similar too.**

**Lets chek the different values is Case number, Case Number.1 and Case Sumber.2**

In [28]:
data[['Date','Year','Sex','Case Number','Case Number.1','Case Number.2']].loc[(data['Case Number'] != data['Case Number.1'])| (data['Case Number'] != data['Case Number.2']) | (data['Case Number.1'] != data['Case Number.2'])]

Date  Year  Sex   Case Number  \
4                                16-Sep-16  2016    M    2016.09.15   
33                   Reported  14-Jul-2016  2016    M  2016.07.14.4   
97                               24-Jan-16  2016    M  2016.01.24.b   
116                              07-Nov-15  2015    F    2015.12.23   
121                              28-Oct-15  2015    M  2015.10.28.a   
169                              10-Jul-15  2015    M    2015.07-10   
3296                             05-Jul-67  1967    M    1967.07.05   
3569                             30-Aug-62  1962    M  1962,08.30.b   
3654                  Reported 06-Sep-1961  1961    M  1961.09.02.R   
4177                             05-Aug-52  1952  NaN    1952.08.05   
4719                  Reported 08-Feb-1934  1934    M  1934.01.08.R   
5043  Reported to have taken place in 1919  1919    M  1900.00.00.R   
5150                  Reported 31-Jul-1911  1911    M  1911.07.31.R   

     Case Number.1 Case Number.2  
4       2016.09.16    2016.09.15  
33    2016.07.14.R  2016.07.14.4  
97    2015.01.24.b  2016.01.24.b  
116     2015.11.07    2015.12.23  
121     2015.10.28  2015.10.28.a  
169     2015.07.10    2015.07.10  
3296    1967/07.05    1967.07.05  
3569  1962.08.30.b  1962,08.30.b  
3654  1961.09,06.R  1961.09.02.R  
4177    1952.08.04    1952.08.05  
4719  1934.02.08.R  1934.02.08.R  
5043  1919.00.00.R  1900.00.00.R  
5150  1911.07.31.T  1911.07.31.R

**First I will make the columns to be type string and then replace all the comas in the Case Number, Case Number.1 and Case Number.2**
**As I guess the rules for the naming are year.month.day.victim.Reported yyyy.mm.dd.[a].[R]**

In [29]:
data[['Case Number', 'Case Number.1', 'Case Number.2']] = data[['Case Number', 'Case Number.1', 'Case Number.2']].astype(str)


In [30]:
data['Case Number']= data['Case Number'].str.replace(',', '.')

In [31]:
data['Case Number.1']= data['Case Number.1'].str.replace(',', '.')

In [32]:
data['Case Number.2']= data['Case Number.2'].str.replace(',', '.')

**I think Case Number.1 and Case Number.2 are corrections about the Case Number**

In [33]:
data[['Date','Year','Sex','Case Number','Case Number.1','Case Number.2']].loc[(data['Case Number'] != data['Case Number.1'])| (data['Case Number'] != data['Case Number.2']) | (data['Case Number.1'] != data['Case Number.2'])]

Date  Year  Sex   Case Number  \
4                                16-Sep-16  2016    M    2016.09.15   
33                   Reported  14-Jul-2016  2016    M  2016.07.14.4   
97                               24-Jan-16  2016    M  2016.01.24.b   
116                              07-Nov-15  2015    F    2015.12.23   
121                              28-Oct-15  2015    M  2015.10.28.a   
169                              10-Jul-15  2015    M    2015.07-10   
3296                             05-Jul-67  1967    M    1967.07.05   
3654                  Reported 06-Sep-1961  1961    M  1961.09.02.R   
4177                             05-Aug-52  1952  NaN    1952.08.05   
4719                  Reported 08-Feb-1934  1934    M  1934.01.08.R   
5043  Reported to have taken place in 1919  1919    M  1900.00.00.R   
5150                  Reported 31-Jul-1911  1911    M  1911.07.31.R   

     Case Number.1 Case Number.2  
4       2016.09.16    2016.09.15  
33    2016.07.14.R  2016.07.14.4  
97    2015.01.24.b  2016.01.24.b  
116     2015.11.07    2015.12.23  
121     2015.10.28  2015.10.28.a  
169     2015.07.10    2015.07.10  
3296    1967/07.05    1967.07.05  
3654  1961.09.06.R  1961.09.02.R  
4177    1952.08.04    1952.08.05  
4719  1934.02.08.R  1934.02.08.R  
5043  1919.00.00.R  1900.00.00.R  
5150  1911.07.31.T  1911.07.31.R

**I pick the correct data from the other columns**

In [34]:
data['Case Number'].iloc[169] = data['Case Number.2'].iloc[169]

/usr/local/lib/python3.7/site-packages/pandas/core/indexing.py:190: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self._setitem_with_indexer(indexer, value)


In [35]:
data['Case Number'].iloc[[4, 33, 116, 121, 3569, 3654, 4719, 5043]] = data['Case Number.1'].iloc[[4, 33, 116, 121, 3569, 3654, 4719, 5043]]


In [36]:
data[['Date','Year','Sex','Case Number','Case Number.1','Case Number.2']].loc[(data['Case Number'] != data['Case Number.1'])| (data['Case Number'] != data['Case Number.2']) | (data['Case Number.1'] != data['Case Number.2'])]

Date  Year  Sex   Case Number  \
4                                16-Sep-16  2016    M    2016.09.16   
33                   Reported  14-Jul-2016  2016    M  2016.07.14.R   
97                               24-Jan-16  2016    M  2016.01.24.b   
116                              07-Nov-15  2015    F    2015.11.07   
121                              28-Oct-15  2015    M    2015.10.28   
3296                             05-Jul-67  1967    M    1967.07.05   
3654                  Reported 06-Sep-1961  1961    M  1961.09.06.R   
4177                             05-Aug-52  1952  NaN    1952.08.05   
5043  Reported to have taken place in 1919  1919    M  1919.00.00.R   
5150                  Reported 31-Jul-1911  1911    M  1911.07.31.R   

     Case Number.1 Case Number.2  
4       2016.09.16    2016.09.15  
33    2016.07.14.R  2016.07.14.4  
97    2015.01.24.b  2016.01.24.b  
116     2015.11.07    2015.12.23  
121     2015.10.28  2015.10.28.a  
3296    1967/07.05    1967.07.05  
3654  1961.09.06.R  1961.09.02.R  
4177    1952.08.04    1952.08.05  
5043  1919.00.00.R  1900.00.00.R  
5150  1911.07.31.T  1911.07.31.R

**Now I can drop the columns Case Number.1 and Case Number.2**

In [37]:
data = data.drop(['Case Number.1', 'Case Number.2'], axis=1)

In [38]:
data.head()

Case Number       Date  Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  13h00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  11h00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  10h43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016    NaN  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016    NaN  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex  Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M   16   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M   36   
2  New Smyrna Beach, Volusia County  Surfing             male   M   43   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M  NaN   
4                       Bells Beach  Surfing             male   M  NaN   

                                  Injury Fatal    Species  \
0                  Minor injury to thigh     N        NaN   
1                   Lacerations to hands     N        NaN   
2               Lacerations to lower leg     N        NaN   
3           Struck by fin on chest & leg     N        NaN   
4  No injury: Knocked off board by shark     N  2 m shark   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                                                      href formula  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

                                                                              href  
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf  
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf  
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf  
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf  
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf

**Lets chek the different values href and href formula**

In [39]:
data[['Date','Year','Sex','href','href formula','Case Number']].loc[data['href'] != data['href formula']]


Date  Year  Sex  \
20               29-Jul-16  2016    M   
27               23-Jul-16  2016    M   
61               21-May-16  2016    F   
107              21-Dec-15  2015    M   
114              15-Nov-15  2015    M   
134              05-Oct-15  2015    M   
180              27-Jun-15  2015    M   
193              13-Jun-15  2015    F   
232              24-Jan-15  2015    M   
262              19-Oct-14  2014    F   
263              18-Oct-14  2014    M   
264              17-Oct-14  2014    F   
271              03-Oct-14  2014    M   
272              03-Oct-14  2014    M   
293              12-Aug-14  2014    M   
305              22-Jul-14  2014    M   
323              09-Jun-14  2014    M   
347              03-Apr-14  2014    F   
361              07-Feb-14  2014    M   
362              07-Feb-14  2014    M   
363              26-Jan-14  2014    M   
364              25-Jan-14  2014    M   
365              04-Jan-14  2014    M   
367              25-Dec-13  2013    M   
377              22-Nov-13  2013    M   
378              12-Nov-13  2013    M   
379              10-Nov-13  2013    M   
380              07-Nov-13  2013    M   
381              07-Nov-13  2013    M   
382              31-Oct-13  2013    M   
383              28-Oct-13  2013    F   
384              26-Oct-13  2013    M   
448              27-May-13  2013    M   
449              27-May-13  2013    M   
823              01-Feb-10  2010    F   
1217             22-Aug-06  2006    M   
1218             22-Aug-06  2006    M   
2274                Apr-93  1993    M   
2477                  1989  1989    M   
3019             19-Jan-75  1975    M   
3549                  1963  1963    M   
3603             25-Mar-62  1962    M   
3917             31-Aug-58  1958  NaN   
3928                Jul-58  1958    M   
4394             30-Jul-45  1945  NaN   
4642  Reported 04-Aug-1936  1936    M   
4668             24-Aug-35  1935    M   
4719  Reported 08-Feb-1934  1934    M   
5317             05-Sep-00  1900    M   
5458            06-26-1890  1890    M   
5686           02-Jan-1864  1864    M   
5694         Reported 1863  1863    M   
5819  Reported 08-Jul-1819  1819    M   
5857         Reported 1637  1637  NaN   

                                                                                                                                         href  \
20           http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2016.07.29-Spain.pdf   
27                                                              http://sharkattackfile.net/spreadsheets/pdf_directory/2016.07.23-Cutbirth.pdf   
61          http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2016.05.21.a-Girl.pdf   
107       http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2015.12.21.a-Brazil.pdf   
114     http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2015.11.15.a-Engelman.pdf   
134                                                            http://sharkattackfile.net/spreadsheets/pdf_directory/2015.10.05-Slaughter.pdf   
180                                                            http://sharkattackfile.net/spreadsheets/pdf_directory/2015.06.27-Swanepoel.pdf   
193                                                                                                                                       NaN   
232         http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2015.01.24-Murray.pdf   
262           http://sharkattackfile.net/spreadsheets/pdf_directory/http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.19-Tara.pdf   
263                                                                 http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.18-Roberson   
264                       

**I assume href formula is the most accurate column for the urls, anyhow I will correct some errors**

**Take some values from href as the cell has the correct url**

In [40]:
data['href formula'].iloc[[193,293,364,448,823,3019]] = data['href'].iloc[[193,293,364,448,823,3019]]

In [41]:
data['href formula'].iloc[363] = None

In [42]:
index_href = data['href formula'].loc[data['href formula'].str.endswith('.pdf') == False].index

In [43]:
data['href formula'].iloc[index_href]

65               http://sharkattackfile.net/spreadsheets/pdf_directory/2016.05.02-NZ-pdf
126      http://sharkattackfile.net/spreadsheets/pdf_directory/2015.10.17.b.-Hawaii. pdf
157           http://sharkattackfile.net/spreadsheets/pdf_directory/2015.08.19-Szsarmack
177          http://sharkattackfile.net/spreadsheets/pdf_directory/2015.06.30.b-Steve.pf
226         http://sharkattackfile.net/spreadsheets/pdf_directory/2015.02.14-Bishop.pdf 
251      http://sharkattackfile.net/spreadsheets/pdf_directory/2014.11.20-Mauritius.pdf 
252           http://sharkattackfile.net/spreadsheets/pdf_directory/2014.11.19-Freo.pdf 
264         http://sharkattackfile.net/spreadsheets/pdf_directory/2014.10.17-Olsson, pdf
295      http://sharkattackfile.net/spreadsheets/pdf_directory/2014.08.09.R-Hindley, pdf
334         http://sharkattackfile.net/spreadsheets/pdf_directory/2014.05.14-McLeod, pdf
343       http://sharkattackfile.net/spreadsheets/pdf_directory/.2014.04.12.R-inflatable
352            http:/

In [44]:
data['href formula'] = data['href formula'].str.replace(', pdf', '.pdf')
data['href formula'] = data['href formula'].str.replace('. Pdf', '.pdf')
data['href formula'] = data['href formula'].str.replace('-pdf', '.pdf')
data['href formula'] = data['href formula'].str.replace('.pf', '.pdf')
data['href formula'] = data['href formula'].str.replace('. pdf', '.pdf')
data['href formula'] = data['href formula'].str.replace(' ','')
data['href formula'] = data['href formula'].str.replace('/ .','')

In [45]:
data['href formula'] = data['href formula'].str.replace('/.2014','/2014')

In [46]:
data['href formula'].loc[data['href formula'].str.endswith('.pdf') == False]

157        http://sharkattackfile.net/spreadsheets/pdf_directory/2015.08.19-Szsarmack
343     http://sharkattackfile.net/spreadsheets/pdf_directory/2014.04.12.R-inflatable
352         http://sharkattackfile.net/spreadsheets/pdf_directory/2014.03.18.c-Borgna
359         http://sharkattackfile.net/spreadsheets/pdf_directory/2014.02.17.R-OneDLL
425       http://sharkattackfile.net/spreadsheets/pdf_directory/2013.07.29.a-Gatewood
1004          http://sharkattackfile.net/spreadsheets/pdf_directory/2008.07.26.a-Maui
2476      http://sharkattackfile.net/spreadsheets/pdf_directory/1989.00.00.b-Halstead
2714        http://sharkattackfile.net/spreadsheets/pdf_directory/1983.06.00-Starling
3987          http://sharkattackfile.net/spreadsheets/pdf_directory/1957.00.00.i-baby
4481       http://sharkattackfile.net/spreadsheets/pdf_directory/1942.07.06-Bardolini
5257                 http://sharkattackfile.net/spreadsheets/pdf_directory/1905.07.00
5329         http://sharkattackfile.net/spreadsheets/p

**The href formula is as complete and accurate I can. I'm going to delete the href column and rename the href formula column**

In [47]:
data = data.drop(['href'], axis=1)

In [48]:
data = data.rename(index=str, columns={"href formula": "url"})

In [49]:
data.head()

Case Number       Date  Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  13h00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  11h00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  10h43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016    NaN  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016    NaN  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex  Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M   16   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M   36   
2  New Smyrna Beach, Volusia County  Surfing             male   M   43   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M  NaN   
4                       Bells Beach  Surfing             male   M  NaN   

                                  Injury Fatal    Species  \
0                  Minor injury to thigh     N        NaN   
1                   Lacerations to hands     N        NaN   
2               Lacerations to lower leg     N        NaN   
3           Struck by fin on chest & leg     N        NaN   
4  No injury: Knocked off board by shark     N  2 m shark   

        Investigator or Source                        pdf  \
0  Orlando Sentinel, 9/19/2016       2016.09.18.c-NSB.pdf   
1  Orlando Sentinel, 9/19/2016   2016.09.18.b-Luciano.pdf   
2  Orlando Sentinel, 9/19/2016       2016.09.18.a-NSB.pdf   
3           The Age, 9/18/2016  2016.09.17-Angiolella.pdf   
4           The Age, 9/16/2016  2016.09.16-BellsBeach.pdf   

                                                                               url  
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf  
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf  
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf  
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf  
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf

**pdf contains the same information than href, so I will delete it**

In [50]:
data = data.drop(['pdf'], axis=1)

In [51]:
data.head()

Case Number       Date  Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  13h00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  11h00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  10h43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016    NaN  Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016    NaN  Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex  Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M   16   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M   36   
2  New Smyrna Beach, Volusia County  Surfing             male   M   43   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M  NaN   
4                       Bells Beach  Surfing             male   M  NaN   

                                  Injury Fatal    Species  \
0                  Minor injury to thigh     N        NaN   
1                   Lacerations to hands     N        NaN   
2               Lacerations to lower leg     N        NaN   
3           Struck by fin on chest & leg     N        NaN   
4  No injury: Knocked off board by shark     N  2 m shark   

        Investigator or Source  \
0  Orlando Sentinel, 9/19/2016   
1  Orlando Sentinel, 9/19/2016   
2  Orlando Sentinel, 9/19/2016   
3           The Age, 9/18/2016   
4           The Age, 9/16/2016   

                                                                               url  
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf  
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf  
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf  
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf  
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf

### 2.1.4 Columns with a lot of Null Values

**Three more columns have a lot of Nulls, we check them: Age, Time, Species.**

**Lests check the different values of Age**

In [52]:
data['Age'].unique()

array(['16', '36', '43', nan, '60s', '51', '50', '12', '9', '22', '25',
       '37', '20', '49', '15', '21', '40', '72', '18', '29', '31', '11',
       '10', '59', '42', '34', '35', '19', '6', '27', '64', '60', '23',
       '52', '13', '57', '48', '39', '24', '26', '69', '46', 'Teen', '41',
       '45', '65', '38', '71', '32', '58', '28', '54', '44', '14', '7',
       '62', '40s', '68', '47', '17', '30', '63', '70', '18 months', '53',
       '20s', '33', '30s', '50s', '8', '61', '55', 'teen', '66', '77',
       '74', '3', '56', '28 & 26', '5', '86', '18 or 20', '12 or 13',
       '46 & 34', '28, 23 & 30', 'Teens', '36 & 26', '8 or 10', '84',
       '\xa0 ', ' ', '30 or 36', '6½', '21 & ?', '75', '33 or 37',
       'mid-30s', '73', '23 & 20', '7      &    31', '20?', "60's",
       '32 & 30', '16 to 18', '87', '67', 'Elderly', 'mid-20s', 'Ca. 33',
       '21 or 26', '>50', '18 to 22', 'adult', '9 & 12', '? & 19',
       '9 months', '25 to 35', '23 & 26', '1', '(adult)', '33 & 37',
     

**I will change all the non numeric values by NaN, pass the Age to numeric, fill the NaN to 0.0 and finally convert all the types to integer**

In [53]:
data['Age'] = pd.to_numeric(data['Age'], errors='coerce') 

In [54]:
data['Age'].unique()

array([16., 36., 43., nan, 51., 50., 12.,  9., 22., 25., 37., 20., 49.,
       15., 21., 40., 72., 18., 29., 31., 11., 10., 59., 42., 34., 35.,
       19.,  6., 27., 64., 60., 23., 52., 13., 57., 48., 39., 24., 26.,
       69., 46., 41., 45., 65., 38., 71., 32., 58., 28., 54., 44., 14.,
        7., 62., 68., 47., 17., 30., 63., 70., 53., 33.,  8., 61., 55.,
       66., 77., 74.,  3., 56.,  5., 86., 84., 75., 73., 87., 67.,  1.,
       81., 78.])

**Lests check the different values of Time**

In [55]:
data['Time'].unique()

array(['13h00', '11h00', '10h43', nan, '15h15', '14h30', '15h40',
       'Late afternoon', '15h00', '14h00', '17h00', '16h00', '11h30',
       '12h00', '19h05', 'After noon', '1300', '10h00', '14h30 / 15h30',
       '22h00', '16h20', '14h34', '15h25', '14h55', '06h00', '17h30',
       '08h30', 'Morning', 'Afternoon', '15h45', '18h00', '17h46',
       'Morning ', '13h20', '15h49', '07h00', '08h00', '10h45', '19h00',
       '13h30', 'Midnight', '12h30', '09h30 / 10h00', '10h30', '18h15',
       '04h00', '14h50', '13h50', '19h20', '10h25', '16h50',
       '10h45-11h15', '16h45', '15h52', '07h30', '06h15', '14h40',
       '19h45', '17h10', '15h30', '10h10', '08h45', 'Evening', '09h00',
       '12h10', '18h05', '19h30', '11h41', '12h25', '09h30', '17h51',
       '16h12', '12h45', '10h50', '13h45', '14h15', '16h30', '09h45',
       '11h45', '10h20', '18h30', '06h45', '05h00', '03h30', 'Night',
       'Sometime between 06h00 & 08hoo', '06h30', '11h15', '17h20',
       '18h45', '20h00', '16h40

In [56]:
data['Time'] = data['Time'].str.replace('h','')
data['Time'] = data['Time'].str.replace('j','')

In [57]:
data['Time'].unique()

array(['1300', '1100', '1043', nan, '1515', '1430', '1540',
       'Late afternoon', '1500', '1400', '1700', '1600', '1130', '1200',
       '1905', 'After noon', '1000', '1430 / 1530', '2200', '1620',
       '1434', '1525', '1455', '0600', '1730', '0830', 'Morning',
       'Afternoon', '1545', '1800', '1746', 'Morning ', '1320', '1549',
       '0700', '0800', '1045', '1900', '1330', 'Midnigt', '1230',
       '0930 / 1000', '1030', '1815', '0400', '1450', '1350', '1920',
       '1025', '1650', '1045-1115', '1645', '1552', '0730', '0615',
       '1440', '1945', '1710', '1530', '1010', '0845', 'Evening', '0900',
       '1210', '1805', '1930', '1141', '1225', '0930', '1751', '1612',
       '1245', '1050', '1345', '1415', '1630', '0945', '1145', '1020',
       '1830', '0645', '0500', '0330', 'Nigt',
       'Sometime between 0600 & 08oo', '0630', '1115', '1720', '1845',
       '2000', '1640', '1618', '1110', 'Early afternoon', '1750',
       '0700 - 0800', '1815-1830', '1701', '1015', '0957'

In [58]:
data['Time'] = pd.to_numeric(data['Time'], errors='coerce') 

In [59]:
data['Time'] = data['Time'].fillna(-1.0)

In [60]:
data['Time'] = data['Time'].astype(int)

In [61]:
data['Time'].unique()

array([ 1300,  1100,  1043,    -1,  1515,  1430,  1540,  1500,  1400,
        1700,  1600,  1130,  1200,  1905,  1000,  2200,  1620,  1434,
        1525,  1455,   600,  1730,   830,  1545,  1800,  1746,  1320,
        1549,   700,   800,  1045,  1900,  1330,  1230,  1030,  1815,
         400,  1450,  1350,  1920,  1025,  1650,  1645,  1552,   730,
         615,  1440,  1945,  1710,  1530,  1010,   845,   900,  1210,
        1805,  1930,  1141,  1225,   930,  1751,  1612,  1245,  1050,
        1345,  1415,  1630,   945,  1145,  1020,  1830,   645,   500,
         330,   630,  1115,  1720,  1845,  2000,  1640,  1618,  1110,
        1750,  1701,  1015,   957,   820,  1758,  2030,  1519,  1055,
        1555,  2300,  1240,  1355,  1605,  1410,  1520,  1250,  1445,
        1340,  1324,   940,  1140,   810,  1556,  1420,   756,  1717,
        1635,   905,   715,   720,  1928,  1238,   550,  1550,  1105,
         545,  1325,   950,  1220,  1326,   911,  1215,  1315,  1745,
        1820,  1040,

In [62]:
data['hour'] = round(data['Time']/100).astype(int)

In [63]:
data['hour'].loc[data['hour'] > 24] = -1

In [64]:
data['hour'].loc[data['hour'] == 0] = '00'
data['hour'].loc[data['hour'] == 1] = '01'
data['hour'].loc[data['hour'] == 2] = '02'
data['hour'].loc[data['hour'] == 3] = '03'
data['hour'].loc[data['hour'] == 4] = '04'
data['hour'].loc[data['hour'] == 5] = '05'
data['hour'].loc[data['hour'] == 6] = '06'
data['hour'].loc[data['hour'] == 7] = '07'
data['hour'].loc[data['hour'] == 8] = '08'
data['hour'].loc[data['hour'] == 9] = '09'

In [65]:
data['hour'].unique()

array([13, 11, 10, '00', 15, 14, 17, 16, 12, 19, 22, '06', '08', 18, '07',
       '04', '09', '05', '03', 20, 23, 21, -1, '02', '01'], dtype=object)

In [66]:
data['min'] = round(data['Time']%100).astype(int)

In [67]:
data['min'].loc[data['min'] > 59] = -1

In [68]:
data['min'].loc[data['min'] == 0] = '00'
data['min'].loc[data['min'] == 1] = '01'
data['min'].loc[data['min'] == 2] = '02'
data['min'].loc[data['min'] == 3] = '03'
data['min'].loc[data['min'] == 4] = '04'
data['min'].loc[data['min'] == 5] = '05'
data['min'].loc[data['min'] == 6] = '06'
data['min'].loc[data['min'] == 7] = '07'
data['min'].loc[data['min'] == 8] = '08'
data['min'].loc[data['min'] == 9] = '09'
data['min'].loc[data['hour'] == -1] = ''

In [69]:
data['min'].unique()

array(['00', 43, -1, 15, 30, 40, '05', 20, 34, 25, 55, 45, 46, 49, 50, 52,
       10, 41, 51, 12, 18, '01', 57, 58, 19, 24, 56, 17, 35, 28, 38, 26,
       11, 44, 27, 16, 21, 53, '', 42, 39, '08', '02', 14, 37, '07', 23,
       47, '06', 48, 33, 54, 22, 32], dtype=object)

In [70]:
data['Time'] = data['hour'].astype(str) + ':' + data['min'].astype(str) 

In [71]:
data['Time'].loc[data['hour'] == -1] = ''
data['Time'].loc[data['min'] == -1] = ''

In [72]:
data['Time'].unique()

array(['13:00', '11:00', '10:43', '', '15:15', '14:30', '15:40', '15:00',
       '14:00', '17:00', '16:00', '11:30', '12:00', '19:05', '10:00',
       '22:00', '16:20', '14:34', '15:25', '15:55', '06:00', '17:30',
       '08:30', '15:45', '18:00', '17:46', '13:20', '15:49', '07:00',
       '08:00', '10:45', '19:00', '13:30', '12:30', '10:30', '18:15',
       '04:00', '14:50', '19:20', '10:25', '16:50', '16:45', '16:52',
       '07:30', '06:15', '14:40', '19:45', '17:10', '15:30', '10:10',
       '08:45', '09:00', '12:10', '18:05', '19:30', '11:41', '12:25',
       '09:30', '18:51', '16:12', '12:45', '10:50', '13:45', '14:15',
       '16:30', '09:45', '11:45', '10:20', '18:30', '06:45', '05:00',
       '03:30', '06:30', '11:15', '17:20', '18:45', '20:00', '16:40',
       '16:18', '11:10', '18:50', '17:01', '10:15', '10:57', '08:20',
       '18:58', '20:30', '15:19', '11:55', '16:55', '23:00', '12:40',
       '14:55', '16:05', '14:10', '15:20', '12:50', '14:45', '13:40',
       '13:24', 

In [73]:
data = data.drop(['min'], axis=1)

In [74]:
data = data.drop(['hour'], axis=1)

In [75]:
data.head()

Case Number       Date  Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  13:00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  11:00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  10:43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016         Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016         Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex   Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M  16.0   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M  36.0   
2  New Smyrna Beach, Volusia County  Surfing             male   M  43.0   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M   NaN   
4                       Bells Beach  Surfing             male   M   NaN   

                                  Injury Fatal    Species  \
0                  Minor injury to thigh     N        NaN   
1                   Lacerations to hands     N        NaN   
2               Lacerations to lower leg     N        NaN   
3           Struck by fin on chest & leg     N        NaN   
4  No injury: Knocked off board by shark     N  2 m shark   

        Investigator or Source  \
0  Orlando Sentinel, 9/19/2016   
1  Orlando Sentinel, 9/19/2016   
2  Orlando Sentinel, 9/19/2016   
3           The Age, 9/18/2016   
4           The Age, 9/16/2016   

                                                                               url  
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf  
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf  
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf  
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf  
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf

**Lests check the different values of species**

In [76]:
species = data['Species'].unique()

In [77]:
for specie in species:
    print(specie)

nan
2 m shark
3' to 4' shark
Tiger shark, 10?
White shark
Bull shark, 6'
Bull shark, 3.5 m
Tiger shark, 10' 
Blacktip shark
Blue shark
Nurse shark, 4'
reef shark, 1m
Lemon shark, 9'
Angel shark
dogfish shark
Tawny nurse shark
Caribbean reef shark, 4.5'
3' to 5' shark
5' shark
3' shark
Said to involve an 8' shark but more likely damage caused by debris
3+ m shark
Mako shark
Bronze whaler sharks x 3
3 m shark
Bull shark, 4' to 5'
Blacktip shark, 4'
Nurse shark, 2'
Mako shark, 1.5 m [5'] 
Wobbegong shark
Grey reef shark, 2 m
Bull shark, 7'
Tiger shark, 2.5 m
Shark involvement not confirmed
5' to 6' shark
Nurse shark, 2.5-ft
Bronze whaler
Sandtiger shark
No shark involvement
Reef shark, 5'
Bull sharks x 2
Tiger shark, 14'
White shark, 3 m
Silky shark, 6.5'
9' shark
Bronze whaler shark, 6'
Tiger shark, 1.5 m 
White shark, 3.5 m
Bull shark
Spinner shark, 7'
Bull shark, 2.8 to 3.1 m
White shark, 3 to 3.5m 
Spinner shark, 5'
Tiger shark, 7'
Hammerhead sp.
Tiger shark, 13' 
4' to 5' shark
Black

12' tiger shark
Blue pointers
Said to involve a grey nurse shark that leapt out of the water and  seized the boy but species identification is questionable


**Species is to diverse, not usefull for the study. I will delete it**

In [78]:
data = data.drop(['Species'], axis=1)

In [79]:
data.head()

Case Number       Date  Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016  13:00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016  11:00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016  10:43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016         Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016         Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex   Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M  16.0   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M  36.0   
2  New Smyrna Beach, Volusia County  Surfing             male   M  43.0   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M   NaN   
4                       Bells Beach  Surfing             male   M   NaN   

                                  Injury Fatal       Investigator or Source  \
0                  Minor injury to thigh     N  Orlando Sentinel, 9/19/2016   
1                   Lacerations to hands     N  Orlando Sentinel, 9/19/2016   
2               Lacerations to lower leg     N  Orlando Sentinel, 9/19/2016   
3           Struck by fin on chest & leg     N           The Age, 9/18/2016   
4  No injury: Knocked off board by shark     N           The Age, 9/16/2016   

                                                                               url  
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf  
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf  
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf  
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf  
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf

**Let's Work with the column YEAR**

In [80]:
data['Year'].unique()

array([2016, 2015, 2014, 2013, 2012, 2011, 2010, 2009, 2008, 2007, 2006,
       2005, 2004, 2003, 2002, 2001, 2000, 1999, 1998, 1997, 1996, 1995,
       1984, 1994, 1993, 1992, 1991, 1990, 1989, 1969, 1988, 1987, 1986,
       1985, 1983, 1982, 1981, 1980, 1979, 1978, 1977, 1976, 1975, 1974,
       1973, 1972, 1971, 1970, 1968, 1967, 1966, 1965, 1964, 1963, 1962,
       1961, 1960, 1959, 1958, 1957, 1956, 1955, 1954, 1953, 1952, 1951,
       1950, 1949, 1948, 1848, 1947, 1946, 1945, 1944, 1943, 1942, 1941,
       1940, 1939, 1938, 1937, 1936, 1935, 1934, 1933, 1932, 1931, 1930,
       1929, 1928, 1927, 1926, 1925, 1924, 1923, 1922, 1921, 1920, 1919,
       1918, 1917, 1916, 1915, 1914, 1913, 1912, 1911, 1910, 1909, 1908,
       1907, 1906, 1905, 1904, 1903, 1902, 1901, 1900, 1899, 1898, 1897,
       1896, 1895, 1894, 1893, 1892, 1891, 1890, 1889, 1888, 1887, 1886,
       1885, 1884, 1883, 1882, 1881, 1880, 1879, 1878, 1877, 1876, 1875,
       1874, 1873, 1872, 1871, 1870, 1869, 1868, 18

In [81]:
data['Year'].loc[data['Year'] == 0] = np.nan

In [82]:
data['Year'].unique()

array([2016., 2015., 2014., 2013., 2012., 2011., 2010., 2009., 2008.,
       2007., 2006., 2005., 2004., 2003., 2002., 2001., 2000., 1999.,
       1998., 1997., 1996., 1995., 1984., 1994., 1993., 1992., 1991.,
       1990., 1989., 1969., 1988., 1987., 1986., 1985., 1983., 1982.,
       1981., 1980., 1979., 1978., 1977., 1976., 1975., 1974., 1973.,
       1972., 1971., 1970., 1968., 1967., 1966., 1965., 1964., 1963.,
       1962., 1961., 1960., 1959., 1958., 1957., 1956., 1955., 1954.,
       1953., 1952., 1951., 1950., 1949., 1948., 1848., 1947., 1946.,
       1945., 1944., 1943., 1942., 1941., 1940., 1939., 1938., 1937.,
       1936., 1935., 1934., 1933., 1932., 1931., 1930., 1929., 1928.,
       1927., 1926., 1925., 1924., 1923., 1922., 1921., 1920., 1919.,
       1918., 1917., 1916., 1915., 1914., 1913., 1912., 1911., 1910.,
       1909., 1908., 1907., 1906., 1905., 1904., 1903., 1902., 1901.,
       1900., 1899., 1898., 1897., 1896., 1895., 1894., 1893., 1892.,
       1891., 1890.,

**It seems there are some outliers of the Year Column. Lets find them**

In [83]:
stats = data.describe().T
stats

count         mean        std  min     25%     50%     75%     max
Year  5868.0  1965.887185  61.730863  5.0  1945.0  1976.0  2004.0  2016.0
Age   3210.0    27.123364  13.719479  1.0    17.0    24.0    35.0    87.0

In [84]:
stats['IQR'] = stats['75%'] - stats['25%']
stats

count         mean        std  min     25%     50%     75%     max  \
Year  5868.0  1965.887185  61.730863  5.0  1945.0  1976.0  2004.0  2016.0   
Age   3210.0    27.123364  13.719479  1.0    17.0    24.0    35.0    87.0   

       IQR  
Year  59.0  
Age   18.0

In [85]:
stats.index[0]

'Year'

In [86]:
outliers = pd.DataFrame(columns=data.columns)

col = stats.index[0]
iqr = stats.at[col,'IQR']
cutoff = iqr * 1.5
lower = stats.at[col,'25%'] - cutoff
upper = stats.at[col,'75%'] + cutoff
results = data[(data[col] < lower) | (data[col] > upper)].copy()
results['Outlier'] = col
outliers = outliers.append(results) 

/usr/local/lib/python3.7/site-packages/pandas/core/frame.py:6692: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  sort=sort)


In [87]:
outliers

Activity  \
4312                                                                                                                                            Swimming   
5723                                                                                                                                                 NaN   
5724                                                                                                                                            Swimming   
5725                                                                                                                                            Swimming   
5726                                                                                                                                            Swimming   
5727                                                                                                                                            Swimming   
5728                                                                                                            ship William Penn grounded & broke apart   
5729                                                                                                                                             Bathing   
5730                                                                                                                     Fell overboard from the Malacca   
5731                                                                                                                     Fell overboard from the Sobella   
5732                                                                                                                                           Swimming    
5733                                                                                                                                           Swimming    
5734                                                                                                                                            Swimming   
5735                                                                                                                                            Swimming   
5736                                                                                                                           Commercial Salvage Diving   
5737                                                                                                                                            Swimming   
5738                                                                                                                                 Fell into the water   
5739                                                                                                                                     Hard hat diving   
5740                                                                                     He was fighting a shark when his boat capsized & he disappeared   
5741                                                                                                                                   Knocked overboard   
5742                                                                                                                                            Swimming   
5743                                                                                                                                             Fishing   
5744                                                                                                                                      Fell overboard   
5745                                                                                                                                      Fell overboard   
5746                                                                                                                                            Swimming   
5747                                                                                    

In [88]:
outliers.index

Index(['4312', '5723', '5724', '5725', '5726', '5727', '5728', '5729', '5730',
       '5731',
       ...
       '5858', '5859', '5860', '5861', '5862', '5863', '5864', '5865', '5866',
       '5867'],
      dtype='object', length=146)

In [89]:
for index in outliers.index:
    data['Year'].loc[data['Year'].index == index] = np.nan

In [90]:
data['Year'].unique()

array([2016., 2015., 2014., 2013., 2012., 2011., 2010., 2009., 2008.,
       2007., 2006., 2005., 2004., 2003., 2002., 2001., 2000., 1999.,
       1998., 1997., 1996., 1995., 1984., 1994., 1993., 1992., 1991.,
       1990., 1989., 1969., 1988., 1987., 1986., 1985., 1983., 1982.,
       1981., 1980., 1979., 1978., 1977., 1976., 1975., 1974., 1973.,
       1972., 1971., 1970., 1968., 1967., 1966., 1965., 1964., 1963.,
       1962., 1961., 1960., 1959., 1958., 1957., 1956., 1955., 1954.,
       1953., 1952., 1951., 1950., 1949., 1948.,   nan, 1947., 1946.,
       1945., 1944., 1943., 1942., 1941., 1940., 1939., 1938., 1937.,
       1936., 1935., 1934., 1933., 1932., 1931., 1930., 1929., 1928.,
       1927., 1926., 1925., 1924., 1923., 1922., 1921., 1920., 1919.,
       1918., 1917., 1916., 1915., 1914., 1913., 1912., 1911., 1910.,
       1909., 1908., 1907., 1906., 1905., 1904., 1903., 1902., 1901.,
       1900., 1899., 1898., 1897., 1896., 1895., 1894., 1893., 1892.,
       1891., 1890.,

In [91]:
data.head()

Case Number       Date    Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016.0  13:00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016.0  11:00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016.0  10:43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex   Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M  16.0   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M  36.0   
2  New Smyrna Beach, Volusia County  Surfing             male   M  43.0   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M   NaN   
4                       Bells Beach  Surfing             male   M   NaN   

                                  Injury Fatal       Investigator or Source  \
0                  Minor injury to thigh     N  Orlando Sentinel, 9/19/2016   
1                   Lacerations to hands     N  Orlando Sentinel, 9/19/2016   
2               Lacerations to lower leg     N  Orlando Sentinel, 9/19/2016   
3           Struck by fin on chest & leg     N           The Age, 9/18/2016   
4  No injury: Knocked off board by shark     N           The Age, 9/16/2016   

                                                                               url  
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf  
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf  
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf  
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf  
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf

### 2.1.5. Clean the rest of the columns

**Column Date**

In [92]:
data['Date'].unique()

array(['18-Sep-16', '17-Sep-16', '16-Sep-16', ..., '1900-1905',
       '1883-1889', '1845-1853'], dtype=object)

**I will separate Reported from atacks dates**

In [93]:
data['Reported'] = data['Date'][data['Date'].str.startswith('Reported') == True]

In [94]:
data['Date'].loc[data['Date'].str.startswith('Reported') == True] = None

In [95]:
data.head(10)

Case Number       Date    Year   Time        Type        Country  \
0  2016.09.18.c  18-Sep-16  2016.0  13:00  Unprovoked            USA   
1  2016.09.18.b  18-Sep-16  2016.0  11:00  Unprovoked            USA   
2  2016.09.18.a  18-Sep-16  2016.0  10:43  Unprovoked            USA   
3    2016.09.17  17-Sep-16  2016.0         Unprovoked      AUSTRALIA   
4    2016.09.16  16-Sep-16  2016.0         Unprovoked      AUSTRALIA   
5  2016.09.15.R  15-Sep-16  2016.0               Boat      AUSTRALIA   
6    2016.09.11  11-Sep-16  2016.0  15:15  Unprovoked            USA   
7    2016.09.07  07-Sep-16  2016.0  14:30  Unprovoked            USA   
8    2016.09.06  06-Sep-16  2016.0  15:40  Unprovoked  NEW CALEDONIA   
9  2016.09.05.b  05-Sep-16  2016.0         Unprovoked            USA   

                Area                                         Location  \
0            Florida                 New Smyrna Beach, Volusia County   
1            Florida                 New Smyrna Beach, Volusia County   
2            Florida                 New Smyrna Beach, Volusia County   
3           Victoria                                 Thirteenth Beach   
4           Victoria                                      Bells Beach   
5  Western Australia                                          Bunbury   
6            Florida                    Ponte Vedra, St. Johns County   
7             Hawaii                                     Makaha, Oahu   
8     North Province                                           Koumac   
9     South Carolina  Kingston Plantation, Myrtle Beach, Horry County   

          Activity                    Name  Sex   Age  \
0          Surfing                    male    M  16.0   
1          Surfing          Chucky Luciano    M  36.0   
2          Surfing                    male    M  43.0   
3          Surfing         Rory Angiolella    M   NaN   
4          Surfing                    male    M   NaN   
5          Fishing  Occupant: Ben Stratton  NaN   NaN   
6           Wading                    male    M   NaN   
7         Swimming                  female    F  51.0   
8     Kite surfing            David Jewell    M  50.0   
9  Boogie boarding          Rylie Williams    F  12.0   

                                       Injury Fatal  \
0                       Minor injury to thigh     N   
1                        Lacerations to hands     N   
2                    Lacerations to lower leg     N   
3                Struck by fin on chest & leg     N   
4       No injury: Knocked off board by shark     N   
5    Shark rammed boat. No injury to occupant     N   
6                         Minor injury to arm     N   
7    Severe lacerations to shoulder & forearm     N   
8                                       FATAL     Y   
9  Lacerations & punctures to lower right leg     N   

        Investigator or Source  \
0  Orlando Sentinel, 9/19/2016   
1  Orlando Sentinel, 9/19/2016   
2  Orlando Sentinel, 9/19/2016   
3           The Age, 9/18/2016   
4           The Age, 9/16/2016   
5   West Australian, 9/15/2016   
6          News4Jax, 9/11/2016   
7    Hawaii News Now, 9/7/2016   
8       TVANouvelles, 9/6/2016   
9            C. Creswell, GSAF   

                                                                               url  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   
5      http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.15.R-boat.pdf   
6  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.11-PonteVedra.pdf   
7        http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.07-Oahu.pdf   
8      http://sharkattac

**Clean separately Date and Report**

In [96]:
data['Date'] = data['Date'].str.replace(' ','')

In [97]:
data['Date ok'] = data['Date'].str.findall('\d\d-\w\w\w-\d\d')

In [98]:
data.head()

Case Number       Date    Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016.0  13:00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016.0  11:00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016.0  10:43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   

                           Location Activity             Name Sex   Age  \
0  New Smyrna Beach, Volusia County  Surfing             male   M  16.0   
1  New Smyrna Beach, Volusia County  Surfing   Chucky Luciano   M  36.0   
2  New Smyrna Beach, Volusia County  Surfing             male   M  43.0   
3                  Thirteenth Beach  Surfing  Rory Angiolella   M   NaN   
4                       Bells Beach  Surfing             male   M   NaN   

                                  Injury Fatal       Investigator or Source  \
0                  Minor injury to thigh     N  Orlando Sentinel, 9/19/2016   
1                   Lacerations to hands     N  Orlando Sentinel, 9/19/2016   
2               Lacerations to lower leg     N  Orlando Sentinel, 9/19/2016   
3           Struck by fin on chest & leg     N           The Age, 9/18/2016   
4  No injury: Knocked off board by shark     N           The Age, 9/16/2016   

                                                                               url  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

  Reported      Date ok  
0      NaN  [18-Sep-16]  
1      NaN  [18-Sep-16]  
2      NaN  [18-Sep-16]  
3      NaN  [17-Sep-16]  
4      NaN  [16-Sep-16]

**Column Type**

In [99]:
data['Type'].unique()

array(['Unprovoked', 'Boat', 'Provoked', 'Invalid', 'Sea Disaster',
       'Boating'], dtype=object)

**Try to change invalid into other categories using Activity**

In [100]:
data['Type'].loc[data['Type'] == 'Boating'] = 'Boat'

In [101]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Sea disaster')] = 'Sea Disaster'

In [102]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Floating')] = 'Sea Disaster'

In [103]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Surfing')] = 'Boat'

In [104]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'SUP')] = 'Boat'

In [105]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Canoeing')] = 'Boat'

In [106]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Fell overboard')] = 'Boat'

In [107]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'boat capsized')] = 'Boat'

In [108]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Swimming')] = 'Unprovoked'

In [109]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Diving')] = 'Unprovoked'

In [110]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Scuba diving')] = 'Unprovoked'

In [111]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Bathing')] = 'Unprovoked'

In [112]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Fishing')] = 'Provoked'

In [113]:
data['Type'].loc[(data['Type'] == 'Invalid') & (data['Activity'] == 'Spearfishing')] = 'Provoked'

In [114]:
data['Type'].unique()

array(['Unprovoked', 'Boat', 'Provoked', 'Invalid', 'Sea Disaster'],
      dtype=object)

**We delete Activity because is not usefull anymore**

In [115]:
data = data.drop(['Activity'], axis=1)

In [116]:
data.head()

Case Number       Date    Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016.0  13:00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016.0  11:00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016.0  10:43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   

                           Location             Name Sex   Age  \
0  New Smyrna Beach, Volusia County             male   M  16.0   
1  New Smyrna Beach, Volusia County   Chucky Luciano   M  36.0   
2  New Smyrna Beach, Volusia County             male   M  43.0   
3                  Thirteenth Beach  Rory Angiolella   M   NaN   
4                       Bells Beach             male   M   NaN   

                                  Injury Fatal       Investigator or Source  \
0                  Minor injury to thigh     N  Orlando Sentinel, 9/19/2016   
1                   Lacerations to hands     N  Orlando Sentinel, 9/19/2016   
2               Lacerations to lower leg     N  Orlando Sentinel, 9/19/2016   
3           Struck by fin on chest & leg     N           The Age, 9/18/2016   
4  No injury: Knocked off board by shark     N           The Age, 9/16/2016   

                                                                               url  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

  Reported      Date ok  
0      NaN  [18-Sep-16]  
1      NaN  [18-Sep-16]  
2      NaN  [18-Sep-16]  
3      NaN  [17-Sep-16]  
4      NaN  [16-Sep-16]

**Clean Country**

In [117]:
data['Country'].unique()

array(['USA', 'AUSTRALIA', 'NEW CALEDONIA', 'REUNION', 'BAHAMAS', 'SPAIN',
       'CHINA', 'JAPAN', 'COLUMBIA', 'SOUTH AFRICA', 'EGYPT',
       'NEW ZEALAND', 'INDONESIA', 'FRENCH POLYNESIA', 'CAPE VERDE',
       'Fiji', 'BRAZIL', 'DOMINICAN REPUBLIC', 'CAYMAN ISLANDS',
       'UNITED ARAB EMIRATES', 'ARUBA', 'MOZAMBIQUE', 'THAILAND', 'FIJI',
       'PUERTO RICO', 'ITALY', 'MEXICO', 'ATLANTIC OCEAN', 'GREECE',
       'MAURITIUS', nan, 'ST. MARTIN', 'FRANCE', 'ECUADOR',
       'PAPUA NEW GUINEA', 'TRINIDAD & TOBAGO', 'KIRIBATI', 'ISRAEL',
       'DIEGO GARCIA', 'TAIWAN', 'JAMAICA', 'PALESTINIAN TERRITORIES',
       'GUAM', 'SEYCHELLES', 'BELIZE', 'PHILIPPINES', 'NIGERIA', 'TONGA',
       'SCOTLAND', 'CANADA', 'CROATIA', 'SAUDI ARABIA', 'CHILE',
       'ANTIGUA', 'KENYA', 'RUSSIA', 'TURKS & CAICOS', 'COSTA RICA',
       'UNITED KINGDOM', 'MALAYSIA', 'UNITED ARAB EMIRATES (UAE)',
       'SAMOA', 'AZORES', 'SOLOMON ISLANDS', 'SOUTH KOREA', 'MALTA',
       'VIETNAM', 'MADAGASCAR', 'PANAMA',

In [118]:
data['Country'] = data['Country'].str.upper()

**Column Location**

In [119]:
data['Location'].unique()

array(['New Smyrna Beach, Volusia County', 'Thirteenth Beach',
       'Bells Beach', ..., 'Ocracoke Inlet', 'Panama Bay 8ºN, 79ºW',
       'Below the English fort, Trincomalee'], dtype=object)

**I delete this column because is too diverse**

In [120]:
data = data.drop(['Location'], axis=1)

In [121]:
data.head()

Case Number       Date    Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016.0  13:00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016.0  11:00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016.0  10:43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   

              Name Sex   Age                                 Injury Fatal  \
0             male   M  16.0                  Minor injury to thigh     N   
1   Chucky Luciano   M  36.0                   Lacerations to hands     N   
2             male   M  43.0               Lacerations to lower leg     N   
3  Rory Angiolella   M   NaN           Struck by fin on chest & leg     N   
4             male   M   NaN  No injury: Knocked off board by shark     N   

        Investigator or Source  \
0  Orlando Sentinel, 9/19/2016   
1  Orlando Sentinel, 9/19/2016   
2  Orlando Sentinel, 9/19/2016   
3           The Age, 9/18/2016   
4           The Age, 9/16/2016   

                                                                               url  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

  Reported      Date ok  
0      NaN  [18-Sep-16]  
1      NaN  [18-Sep-16]  
2      NaN  [18-Sep-16]  
3      NaN  [17-Sep-16]  
4      NaN  [16-Sep-16]

**Column Sex**

In [122]:
data['Sex'].unique()

array(['M', nan, 'F', 'M ', 'lli', 'N', '.'], dtype=object)

In [123]:
data['Sex'].loc[data['Sex'] == 'M '] = 'M'
data['Sex'].loc[data['Sex'] == 'lli'] = np.nan
data['Sex'].loc[data['Sex'] == 'N'] = np.nan
data['Sex'].loc[data['Sex'] == '.'] = np.nan

In [124]:
data['Sex'].unique()

array(['M', nan, 'F'], dtype=object)

**Column Name**

**I delete the name beacause is not usefull for the study**

In [125]:
data = data.drop(['Name'], axis=1)

In [126]:
data.head()

Case Number       Date    Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016.0  13:00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016.0  11:00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016.0  10:43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   

  Sex   Age                                 Injury Fatal  \
0   M  16.0                  Minor injury to thigh     N   
1   M  36.0                   Lacerations to hands     N   
2   M  43.0               Lacerations to lower leg     N   
3   M   NaN           Struck by fin on chest & leg     N   
4   M   NaN  No injury: Knocked off board by shark     N   

        Investigator or Source  \
0  Orlando Sentinel, 9/19/2016   
1  Orlando Sentinel, 9/19/2016   
2  Orlando Sentinel, 9/19/2016   
3           The Age, 9/18/2016   
4           The Age, 9/16/2016   

                                                                               url  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

  Reported      Date ok  
0      NaN  [18-Sep-16]  
1      NaN  [18-Sep-16]  
2      NaN  [18-Sep-16]  
3      NaN  [17-Sep-16]  
4      NaN  [16-Sep-16]

**Column Fatal(Y/N)**

In [127]:
data['Fatal'].unique()

array(['N', 'Y', nan, 'UNKNOWN', ' N', 'F', 'N ', '#VALUE!', 'n'],
      dtype=object)

In [128]:
data['Fatal'].loc[data['Fatal'] == ' N'] = 'N'
data['Fatal'].loc[data['Fatal'] == 'N '] = 'N'
data['Fatal'].loc[data['Fatal'] == 'F'] = 'Y'
data['Fatal'].loc[data['Fatal'] == 'n'] = 'N'
data['Fatal'].loc[data['Fatal'] == '#VALUE!'] = np.nan

In [129]:
data[['Fatal','Injury']].loc[(data['Fatal'] != 'Y') & (data['Fatal'] != 'N')& (data['Fatal'] != 'UNKNOWN')]

Fatal  \
54     NaN   
1844   NaN   
2449   NaN   
3280   NaN   
3435   NaN   
3901   NaN   
4107   NaN   
4112   NaN   
5307   NaN   
5437   NaN   
5461   NaN   
5468   NaN   
5642   NaN   
5699   NaN   
5718   NaN   
5747   NaN   
5793   NaN   
5794   NaN   
5820   NaN   
5849   NaN   

                                                                                                        Injury  
54                                                    No injury, but sharks repeatedly hit their fins and guns  
1844                                                           Reported as shark attack but probable drowning   
2449                                                                                                     FATAL  
3280         Diver shot the shark, then it injured his arm and broke his leg with its tail.  PROVOKED INCIDENT  
3435         Disappeared, probable drowning but sharks in area led to speculation that he was taken by a shark  
3901                                                                                              Boat damaged  
4107                                No injury to occupants. Shark tore nets & trawl and struck boat repeatedly  
4112                                                                              Human remains found in shark  
5307                                                            Disappeared, but shark involvement unconfirmed  
5437                                                                                      No injury, no attack  
5461  Forensic evidence indicated death resulted from drowning, his body was subsequently scavenged by a shark  
5468                                                                   Human remains found in 4m, 900 kg shark  
5642                                                                                                 No injury  
5699                                                                          Possible drowning and scavenging  
5718                                                                                                     FATAL  
5747                                                                          Death preceded shark involvement  
5793                                                                      Shark caught contained human remains  
5794                                                                     Shark caught, contained human remains  
5820                                                                                        Probable drowning   
5849                            Partial hominid remains recovered from shark, probable drowning and scavenging

In [130]:
data['Fatal'].loc[data['Injury'] == 'FATAL'] = 'Y'

In [131]:
data['Fatal'].unique()

array(['N', 'Y', nan, 'UNKNOWN'], dtype=object)

**Column Injury**

**I delete the name beacause is not usefull anymore for the study**

In [132]:
data = data.drop(['Injury'], axis=1)

In [133]:
data.head()

Case Number       Date    Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016.0  13:00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016.0  11:00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016.0  10:43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   

  Sex   Age Fatal       Investigator or Source  \
0   M  16.0     N  Orlando Sentinel, 9/19/2016   
1   M  36.0     N  Orlando Sentinel, 9/19/2016   
2   M  43.0     N  Orlando Sentinel, 9/19/2016   
3   M   NaN     N           The Age, 9/18/2016   
4   M   NaN     N           The Age, 9/16/2016   

                                                                               url  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

  Reported      Date ok  
0      NaN  [18-Sep-16]  
1      NaN  [18-Sep-16]  
2      NaN  [18-Sep-16]  
3      NaN  [17-Sep-16]  
4      NaN  [16-Sep-16]

In [134]:
Source = data['Investigator or Source'].str.split(",",expand=True)


In [135]:
Source.head()

0           1     2     3     4     5     6     7     8  \
0  Orlando Sentinel   9/19/2016  None  None  None  None  None  None  None   
1  Orlando Sentinel   9/19/2016  None  None  None  None  None  None  None   
2  Orlando Sentinel   9/19/2016  None  None  None  None  None  None  None   
3           The Age   9/18/2016  None  None  None  None  None  None  None   
4           The Age   9/16/2016  None  None  None  None  None  None  None   

      9  
0  None  
1  None  
2  None  
3  None  
4  None

In [136]:
data['Investigator or Source'] = Source[0]

In [137]:
data.head()

Case Number       Date    Year   Time        Type    Country      Area  \
0  2016.09.18.c  18-Sep-16  2016.0  13:00  Unprovoked        USA   Florida   
1  2016.09.18.b  18-Sep-16  2016.0  11:00  Unprovoked        USA   Florida   
2  2016.09.18.a  18-Sep-16  2016.0  10:43  Unprovoked        USA   Florida   
3    2016.09.17  17-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   
4    2016.09.16  16-Sep-16  2016.0         Unprovoked  AUSTRALIA  Victoria   

  Sex   Age Fatal Investigator or Source  \
0   M  16.0     N       Orlando Sentinel   
1   M  36.0     N       Orlando Sentinel   
2   M  43.0     N       Orlando Sentinel   
3   M   NaN     N                The Age   
4   M   NaN     N                The Age   

                                                                               url  \
0       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.c-NSB.pdf   
1   http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.b-Luciano.pdf   
2       http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.18.a-NSB.pdf   
3  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.17-Angiolella.pdf   
4  http://sharkattackfile.net/spreadsheets/pdf_directory/2016.09.16-BellsBeach.pdf   

  Reported      Date ok  
0      NaN  [18-Sep-16]  
1      NaN  [18-Sep-16]  
2      NaN  [18-Sep-16]  
3      NaN  [17-Sep-16]  
4      NaN  [16-Sep-16]

## 2.2. We continue with the rows


### Check the duplicates rows

In [138]:
data.columns

Index(['Case Number', 'Date', 'Year', 'Time', 'Type', 'Country', 'Area', 'Sex',
       'Age', 'Fatal', 'Investigator or Source', 'url', 'Reported', 'Date ok'],
      dtype='object')

In [139]:
duplicate_rows = data[data.duplicated(['Case Number', 'Date', 'Year', 'Type', 'Country', 'Area', 'Sex',
       'Age', 'Fatal', 'Investigator or Source'])]

In [140]:
duplicate_rows

Empty DataFrame
Columns: [Case Number, Date, Year, Time, Type, Country, Area, Sex, Age, Fatal, Investigator or Source, url, Reported, Date ok]
Index: []

**I do not find any duplicated row**

# 3. Final DataFrame

In [141]:
data

Case Number                                                    Date  \
0     2016.09.18.c                                               18-Sep-16   
1     2016.09.18.b                                               18-Sep-16   
2     2016.09.18.a                                               18-Sep-16   
3       2016.09.17                                               17-Sep-16   
4       2016.09.16                                               16-Sep-16   
5     2016.09.15.R                                               15-Sep-16   
6       2016.09.11                                               11-Sep-16   
7       2016.09.07                                               07-Sep-16   
8       2016.09.06                                               06-Sep-16   
9     2016.09.05.b                                               05-Sep-16   
10    2016.09.05.a                                               05-Sep-16   
11      2016.09.04                                               04-Sep-16   
12      2016.09.01                                               01-Sep-16   
13    2016.08.29.b                                               29-Aug-16   
14    2016.08.29.a                                               29-Aug-16   
15      2016.08.27                                               27-Aug-16   
16      2016.08.25                                               25-Aug-16   
17      2016.08.07                                               07-Aug-16   
18      2016.08.06                                               06-Aug-16   
19      2016.08.04                                               04-Aug-16   
20      2016.07.29                                               29-Jul-16   
21    2016.07.28.R                                               28-Jul-16   
22      2016.07.28                                               28-Jul-16   
23      2016.07.27                                               27-Jul-16   
24      2016.07.26                                               26-Jul-16   
25      2016.07.24                                               24-Jul-16   
26    2016.07.23.b                                               23-Jul-16   
27    2016.07.23.a                                               23-Jul-16   
28      2016.07.20                                               20-Jul-16   
29      2016.07.17                                               17-Jul-16   
...            ...                                                     ...   
5962       ND.0031                                              WorldWarII   
5963       ND.0030                                              Before1905   
5964       ND.0028                                     Afewyearsbefore1938   
5965       ND.0027                                                  Nodate   
5966       ND.0026                                              Early1930s   
5967       ND.0025                                              Before1927   
5968       ND.0024                                        Between1918&1939   
5969       ND.0023                                                  Nodate   
5970       ND.0022                                                  Nodate   
5971       ND.0021                                                  Nodate   
5972       ND.0020                                               1920-1923   
5973       ND.0019                                              Before1921   
5974       ND.0018                                              Before1911   
5975       ND.0017                                              Before1921   
5976       ND.0016                                              Before1921   
5977       ND.0015                                              Before1917   
5978       ND.0014                                       Before17-Jul-1916   
5979       ND.0013  Nodate(3daysafterprecedingincident)&priorto19-Jul-1913   
5980       ND.0012                                       Before19-Jul-1913   
5981       ND.0011           

In [143]:
data.to_csv('final_data.csv')